### DISCUSSION TITLE
ALPN Protocol Selection: Bug or Feature?

### DISCUSSION THREAD

**Discussion by mesibo** [openssl/openssl#25914] [Q&A]
We're using `SSL_CTX_set_alpn_select_cb` to select the protocol via ALPN. The client sends a list of mesibo protocols, such as `v1`, `v2`, etc, and the server is expected to choose one from the client’s offerings. Interestingly, we tried setting `v2-xyz` instead of `v2`, and it worked.

While this behavior potentially opens up some optimization opportunities, it’s unclear whether this is a bug that may be fixed in future versions or an intended feature that will remain supported.

According to the `SSL_CTX_set_alpn_select_cb` man page:

> The value of the out, outlen vector should be set to the value of a single protocol selected from the in, inlen vector.

RFC 7301 also specifies, in the context of protocols `advertised by the client`:

> It is expected that a server will have a list of protocols it supports, ordered by preference, and will only select a protocol if the client supports it. In such cases, the server SHOULD select the most preferred protocol it supports that is also advertised by the client.

Could you please clarify if this behavior is intentional and whether we can rely on it in future versions?

**Comment by nhorman** [comment]
RFC 7301 also Indicates that a server SHALL respond with an alpn that is supported by the server and also appears in the client list advertised.

So to respond with an alpn that does not appear in the client list is definitely incorrect.  As to weather or not that should be construed as a bug in the openssl library (ie a failure to validate the outcome vector produced by the alpn call back) or a bug in the application is debatable.

In either case it's definitely not recommended that you rely on the behavior you observe.  You should always return a value from the in vector in the callback

**Reply by davidben** [reply]
On the receiving side, BoringSSL clients will, by default, enforce this rule and fail the connection of the server picks an unsolicited protocol. We do this so the calling application does not get into a bad state because they forget to check it. (Though we did end up needing an option to disable this check because an external user reported that they relied on it, despite the clear wording in the specification.)

**Reply by mesibo** [reply]
We understand that returning a protocol not on the client’s list goes against RFC 7301 and is not recommended. To clarify, this is certainly not an application bug on our end—we were simply testing whether it was possible to send additional parameters along with the response for potential optimizations and to reduce a possible round trip, and it happened to work.

We intended to explore this as an optimization feature only if it was officially supported but we won’t rely on this behavior (unless we maintain a private version), but it would indeed be a useful feature if officially supported in future versions.

**Reply by nhorman** [reply]
It seems like the better solution there might be, rather than augmenting the alpn value, instead register a custom tls extension

**Reply by mesibo** [reply]
> It seems like the better solution there might be, rather than augmenting the alpn value, instead register a custom tls extension

Thanks, Neil, you are right, possibly tls ext would be a good idea.
